### PR TITLE
Update Data Transforms

### LINKED ISSUE #9476: Obsolete
Is there an up-to-date version of this since literally everything on this page is marked "Obsolete"?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a4185570-21cc-dcdb-62fd-d05bceb62f81
* Version Independent ID: 4f0fe52a-a771-d044-593d-2bb65df15e0d
* Content: [Data transforms in ML.NET](https://docs.microsoft.com/en-us/dotnet/machine-learning/resources/transforms)
* Content Source: [docs/machine-learning/resources/transforms.md](https://github.com/dotnet/docs/blob/master/docs/machine-learning/resources/transforms.md)
* Product: **dotnet-ml**
* GitHub Login: @JRAlexander
* Microsoft Alias: **johalex**

### PR DESCRIPTION
**Pull request by JRAlexander**
Update Data Transforms for latest release.

[Internal Review URL](https://review.docs.microsoft.com/en-us/dotnet/machine-learning/resources/transforms?branch=pr-en-us-9371)

Fixes https://github.com/dotnet/docs/issues/9476, https://github.com/dotnet/docs/issues/9552

### CODE REVIEW

File: docs/machine-learning/resources/transforms.md [hunk 1]
```diff
@@ -1,136 +1,185 @@
 ---
-title: Data transforms in ML.NET
```
Comment (justinormont) [inline docs/machine-learning/resources/transforms.md]: Also, I'd reorganize the doc into the name space..

Eg: 
`OneHotEncodingTransformer` should go in the section for `Microsoft.ML.Transforms.Categorical`, "Categorical".

File: docs/machine-learning/resources/transforms.md [hunk 2]
```diff
@@ -1,136 +1,185 @@
...
+title: Machine learning data transforms - ML.NET
 description: Explore the different data transforms supported in ML.NET.
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: Great suggestion!

File: docs/machine-learning/resources/transforms.md [hunk 3]
```diff
@@ -1,136 +1,185 @@
...
-| --- | --- |
-| <xref:Microsoft.ML.Legacy.Transforms.CategoricalHashOneHotVectorizer> | Encodes the categorical variable with hash-based encoding. |
-| <xref:Microsoft.ML.Legacy.Transforms.CategoricalOneHotVectorizer> | Encodes the categorical variable with one-hot encoding based on a term dictionary. |
-
 ## Combiners and segregators
 
 | Transform | Definition |
 | --- | --- |
-| <xref:Microsoft.ML.Legacy.Transforms.CombinerByContiguousGroupId> | Groups values of a scalar column into a vector based on a contiguous group ID. |
+| <xref:Microsoft.ML.Transforms.GroupTransform> | Groups values of a scalar column into a vector based on a contiguous group ID. |
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: Yes, that should be filed as an issue on the dotnet/machinelearning repo as the API ref is auto-generated from the code XMLcomments.

File: docs/machine-learning/resources/transforms.md [hunk 4]
```diff
@@ -1,136 +1,185 @@
...
+| <xref:Microsoft.ML.Transforms.ResNet18Extension> | This is an extension method to be used with the <xref:Microsoft.ML.Transforms.DnnImageFeaturizerEstimator> to use a pretrained ResNet18 model. The NuGet containing this extension is also guaranteed to include the binary model file. |
+| <xref:Microsoft.ML.Transforms.ResNet50Extension> | This is an extension method to be used with the <xref:Microsoft.ML.Transforms.DnnImageFeaturizerEstimator> to use a pretrained ResNet50model. The NuGet containing this extension is also guaranteed to include the binary model file. |
+| <xref:Microsoft.ML.Transforms.ResNet101Extension> | This is an extension method to be used with the <xref:Microsoft.ML.Transforms.DnnImageFeaturizerEstimator> to use a pretrained ResNet101 model. The NuGet containing this extension is also guaranteed to include the binary model file. |
 
 ## Label parsing
 
 | Transform | Definition |
 | --- | --- |
 | <xref:Microsoft.ML.Legacy.Transforms.Dictionarizer> | Converts input values (words, numbers, etc.) to index in a dictionary. |
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: Not that I'm aware of. @asthana86, @shauheen, @shmoradims, or @CESARDELATORRE - any thoughts?

[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: If not, I'll remove it.

File: docs/machine-learning/resources/transforms.md [hunk 5]
```diff
@@ -1,136 +1,185 @@
...
+| <xref:Microsoft.ML.Runtime.ImageAnalytics.ImageLoaderTransform> | Takes one or more ReadOnlyMemory columns and loads them as an ImageType. |
+| <xref:Microsoft.ML.Runtime.ImageAnalytics.ImagePixelExtractorTransform> | Takes one or more ImageType columns and converts them into a vector representation.|
+| <xref:Microsoft.ML.Runtime.ImageAnalytics.ImageResizerTransform> | Takes one or more ImageType columns and resizes them to  the provided height and width.|
+| <xref:Microsoft.ML.Transforms.Text.LatentDirichletAllocationTransformer> | Implements LightLDA, a state-of-the-art implementation of Latent Dirichlet Allocation.|
+| <xref:Microsoft.ML.Transforms.LoadTransform> | Loads specific transforms from the specified model file. Allows for 'cherry picking' transforms from a serialized chain, or to apply a pre-trained transform to a different (but still compatible) data view. |
+| <xref:Microsoft.ML.Transforms.Text.NgramExtractingTransformer> | Produces a bag of counts of ngrams (sequences of consecutive values of length 1-n) in a given vector of keys. It does so by building a dictionary of ngrams and using the id in the dictionary as the index in the bag. |
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: All three or just the last one?

File: docs/machine-learning/resources/transforms.md [hunk 6]
```diff
@@ -1,136 +1,185 @@
 ---
-title: Data transforms in ML.NET
+title: Machine learning data transforms - ML.NET
```
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: But are transforms estimators? That seems confusing.

File: docs/machine-learning/resources/transforms.md [hunk 7]
```diff
@@ -1,136 +1,185 @@
...
-* [Feature selection](#feature-selection)
-* [Featurizers](#featurizers)
-* [Label parsing](#label-parsing)
-* [Missing Values](#missing-values)
-* [Normalization](#normalization)
-* [Row filters](#row-filters)
-* [Schema](#schema)
-* [Text processing and featurization](#text-processing-and-featurization)
-* [Miscellaneous](#miscellaneous)
+# Machine learning data transforms - ML.NET
```
Comment (mairaw) [inline docs/machine-learning/resources/transforms.md]: I find it a bit odd to have the - showing up in H1. I'd make the H1 different here and not use -

Maybe the original H1 was good and the non-branded terms are already added to the title metadata for SEO purposes.

File: docs/machine-learning/resources/transforms.md [hunk 8]
```diff
@@ -1,136 +1,185 @@
...
+| <xref:Microsoft.ML.Transforms.Conversions.HashingTransformer> | Hashes either single valued columns or vector columns. For vector columns, it hashes each slot separately. It can hash either text values or key values. |
+| <xref:Microsoft.ML.Legacy.Transforms.HashConverter> | Converts column values into hashes. This transform accepts both numeric and text inputs, both single and vector-valued columns. |
+| <xref:Microsoft.ML.Transforms.Conversions.HashJoiningTransform> | Converts multiple column values into hashes. This transform accepts both numeric and text inputs, both single and vector-valued columns. |
+| <xref:Microsoft.ML.Transforms.Conversions.KeyToBinaryVectorMappingTransformer> | Converts a key to a binary vector column. |
+| <xref:Microsoft.ML.Transforms.Conversions.KeyToValueMappingTransformer > | Utilizes KeyValues metadata to map key indices to the corresponding values in the KeyValues metadata. |
+| <xref:Microsoft.ML.Transforms.Conversions.KeyToVectorMappingTransformer> | Converts a key to a vector column. |
+| <xref:Microsoft.ML.Transforms.Conversions.TypeConvertingTransformer> | Changes underlying column type provided the type is able to be converted. |
```
Comment (mairaw) [inline docs/machine-learning/resources/transforms.md]: is able to -> can?

File: docs/machine-learning/resources/transforms.md [hunk 9]
```diff
@@ -1,136 +1,185 @@
...
 
 ## Label parsing
 
 | Transform | Definition |
 | --- | --- |
 | <xref:Microsoft.ML.Legacy.Transforms.Dictionarizer> | Converts input values (words, numbers, etc.) to index in a dictionary. |
 | <xref:Microsoft.ML.Legacy.Transforms.LabelColumnKeyBooleanConverter> | Transforms the label to either key or bool (if needed) to make it suitable for classification. |
-| <xref:Microsoft.ML.Legacy.Transforms.LabelIndicator> | Label remapper used by OVA. |
+| <xref:Microsoft.ML.Transforms.LabelConvertTransform> |  Converts labels. |
+| <xref:Microsoft.ML.Transforms.LabelIndicatorTransform> | Remaps multiclass labels to binary T,F labels, primarily for use with OVA.|
```
Comment (mairaw) [inline docs/machine-learning/resources/transforms.md]: so people know what T,F labels are?

File: docs/machine-learning/resources/transforms.md [hunk 10]
```diff
@@ -1,136 +1,185 @@
...
+| --- | --- |
+| <xref:Microsoft.ML.Transforms.OnnxTransform> | Scores pre-trained ONNX models which use the ONNX standard v1.2 |
+
+## Preprocessing
+| Transform | Definition |
+| --- | --- |
+| <xref:Microsoft.ML.Transforms.BootstrapSamplingTransformer> | Approximates bootstrap sampling using Poisson sampling. |
+| <xref:Microsoft.ML.Transforms.Projections.RandomFourierFeaturizingTransformer> | Produces random Fourier feature. |
+| <xref:Microsoft.ML.Transforms.Text.TokenizingByCharactersTransformer> | Character-oriented tokenizer where text is considered a sequence of characters. |
+| <xref:Microsoft.ML.Transforms.Projections.VectorWhiteningTransformer> | Simplfies optimatization to assist with identifying weights. |
```
Comment (mairaw) [inline docs/machine-learning/resources/transforms.md]:
```suggestion
| <xref:Microsoft.ML.Transforms.Projections.VectorWhiteningTransformer> | Simplifies optimization to assist with identifying weights. |
```

File: docs/machine-learning/resources/transforms.md [hunk 11]
```diff
@@ -1,136 +1,185 @@
...
-| <xref:Microsoft.ML.Legacy.Transforms.WordTokenizer> | The input to this transform is text, and the output is a vector of text containing the words (tokens) in the original text. The separator is space, but any other character (or multiple characters) can be specified. |
+| <xref:Microsoft.ML.Transforms.Text.TextNormalizingTransformer> | A text normalization transform that allows normalizing text case, removing diacritical marks, punctuation marks and/or numbers. The transform operates on text input as well as vector of tokens/text (vector of ReadOnlyMemory). |
+| <xref:Microsoft.ML.Transforms.Text.TokenizingByCharactersTransformer> | Character-oriented tokenizer where text is considered a sequence of characters. |
+
+## Time series
+
+| Transform | Definition |
+| --- | --- |
+| <xref:Microsoft.ML.Runtime.TimeSeriesProcessing.ExponentialAverageTransform> | Takes a weighted average of the values: ExpAvg(y_t) = a * y_t + (1-a) * ExpAvg(y_(t-1)). |
+| <xref:Microsoft.ML.Runtime.TimeSeriesProcessing.IidChangePointDetector> | Implements the change point detector transform for an i.i.d. sequence based on adaptive kernel density estimation and martingales. |
```
Comment (mairaw) [inline docs/machine-learning/resources/transforms.md]: do people know what i.i.d. means?

File: docs/machine-learning/resources/transforms.md [hunk 12]
```diff
@@ -1,136 +1,185 @@
...
-| <xref:Microsoft.ML.Legacy.Transforms.DatasetTransformScorer> | Scores a dataset with a transform model. |
-| <xref:Microsoft.ML.Legacy.Transforms.NoOperation> | Does nothing. |
-| <xref:Microsoft.ML.Legacy.Transforms.RandomNumberGenerator> | Adds a column with a generated number sequence. |
-| <xref:Microsoft.ML.Legacy.Transforms.ScoreColumnSelector> | Selects only the last score columns and the extra columns specified in the arguments. |
-| <xref:Microsoft.ML.Legacy.Transforms.Scorer> | Turns the predictor model into a transform model. |
-| <xref:Microsoft.ML.Legacy.Transforms.SentimentAnalyzer> | Uses a pretrained sentiment model to score input strings. |
-| <xref:Microsoft.ML.Legacy.Transforms.TrainTestDatasetSplitter> | Splits the dataset into train and test sets. |
+| <xref:Microsoft.ML.Transforms.CompositeTransformer> | Creates a Composite DataTransform. |
+| <xref:Microsoft.ML.Transforms.CustomMappingTransformer%602> | Generates additional columns to the provided IDataView. It doesn't change the number of rows, and can be seen as a result of application of the user's function to every row of the input data.|
```
Comment (mairaw) [inline docs/machine-learning/resources/transforms.md]: code fence API names or use xref

```suggestion
| <xref:Microsoft.ML.Transforms.CustomMappingTransformer%602> | Generates additional columns to the provided IDataView. It doesn't change the number of rows and can be seen as a result of application of the user's function to every row of the input data.|
```

**Review by mairaw — COMMENTED** [review]
Some quick comments.

File: docs/machine-learning/resources/transforms.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: Removed.

File: docs/machine-learning/resources/transforms.md (hunk 11 above)
Comment (JRAlexander) [reply docs/machine-learning/resources/transforms.md]: probably not. updated.

**Review by mairaw — APPROVED** [review]

**Comment by JRAlexander** [comment]
Thanks, @mairaw!

**Pull request merged**